metricbeat/scripts/mage/docs_collector.go

```diff
@@ -0,0 +1,413 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mage
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	"text/template"
+
+	"github.com/elastic/beats/dev-tools/mage"
+)
+
+// moduleData provides module-level data that will be used to populate the module list
+type moduleData struct {
+	Path       string
+	Base       string
+	Title      string `yaml:"title"`
+	Release    string `yaml:"release"`
+	Dashboards bool
+	Settings   []string `yaml:"settings"`
+	CfgFile    string
+	Asciidoc   string
+	Metricsets []metricsetData
+}
+
+type metricsetData struct {
+	Path       string
+	Title      string
+	Link       string
+	Release    string
+	DataExists bool
+}
+
+func writeTemplate(filename string, t *template.Template, args interface{}) error {
+	fd, err := os.Create(filename)
+	if err != nil {
+		return errors.Wrapf(err, "error opening file at %s", filename)
+	}
+	defer fd.Close()
+	err = t.Execute(fd, args)
+	if err != nil {
+		return errors.Wrap(err, "error executing template")
+	}
+	return nil
+}
+
+//a helper function used by the tempate engine to generate the base paths
+// We're doing this because the mage.*Dir() functions will return an absolute path, which we can't just throw into the docs.
+var funcMap = template.FuncMap{
+	"basePath": func(path string) string {
+		base := "module"
+		if strings.Contains(path, mage.XPackBeatDir()) {
+			base = "../x-pack/metricbeat/module"
+		}
+		return base
+	},
+}
+
+// setupDirectory clears and re-creates the docs/modules directory.
+func setupDirectory() error {
+	docpath := mage.OSSBeatDir("docs/modules")
+	err := os.RemoveAll(docpath)
+	if err != nil {
+		return err
+	}
+	return os.MkdirAll(docpath, 0744)
```

Comment:
Execution permissions as it is a dir?
```suggestion
	return os.MkdirAll(docpath, 0755)
```